File: internal/units/timezone.go
package units

import (
	"fmt"
	"strings"
	"time"
)

// CommonTimezones contains a curated list of commonly used timezones from the tz database
// Each timezone represents a unique STD/DST offset pair to avoid redundancy
// These are verified to exist in the system's tz database
// Total unique offset pairs: 55 (covering all possible timezones globally)
// Ordered from west to east: -11:00 (Niue) to +14:00 (Kiritimati)
var CommonTimezones = []string{
	"Pacific/Niue",           // -11:00
	"America/Adak",           // -10:00/-09:00
	"Pacific/Honolulu",       // -10:00
	"Pacific/Marquesas",      // -09:30
	"America/Anchorage",      // -09:00/-08:00
	"Pacific/Gambier",        // -09:00
	"America/Los_Angeles",    // -08:00/-07:00
	"Pacific/Pitcairn",       // -08:00
	"America/Denver",         // -07:00/-06:00
	"America/Phoenix",        // -07:00
	"America/Chicago",        // -06:00/-05:00
	"America/Mexico_City",    // -06:00
	"America/New_York",       // -05:00/-04:00
	"America/Lima",           // -05:00
	"America/Barbados",       // -04:00
	"America/Santiago",       // -04:00/-03:00
	"America/St_Johns",       // -03:30/-02:30
	"America/Miquelon",       // -03:00/-02:00
	"America/Sao_Paulo",      // -03:00
	"America/Godthab",        // -02:00/-01:00
	"Atlantic/South_Georgia", // -02:00
	"Atlantic/Azores",        // -01:00/+00:00
	"Atlantic/Cape_Verde",    // -01:00
	"UTC",                    // +00:00
	"Africa/Abidjan",         // +00:00
	"Europe/Dublin",          // +00:00/+01:00
	"Antarctica/Troll",       // +00:00/+02:00
	"Africa/Lagos",           // +01:00
	"Europe/Berlin",          // +01:00/+02:00
	"Africa/Johannesburg",    // +02:00
	"Europe/Athens",          // +02:00/+03:00
	"Africa/Nairobi",         // +03:00
	"Asia/Tehran",            // +03:30
	"Asia/Dubai",             // +04:00
	"Asia/Kabul",             // +04:30
	"Asia/Karachi",           // +05:00
	"Asia/Kolkata",           // +05:30
	"Asia/Kathmandu",         // +05:45
	"Asia/Dhaka",             // +06:00
	"Asia/Yangon",            // +06:30
	"Asia/Bangkok",           // +07:00
	"Asia/Singapore",         // +08:00
	"Australia/Eucla",        // +08:45
	"Asia/Seoul",             // +09:00
	"Australia/Darwin",       // +09:30
	"Australia/Adelaide",     // +09:30/+10:30
	"Australia/Brisbane",     // +10:00
	"Australia/Sydney",       // +10:00/+11:00
	"Australia/Lord_Howe",    // +10:30/+11:00
	"Pacific/Bougainville",   // +11:00
	"Pacific/Norfolk",        // +11:00/+12:00
	"Pacific/Fiji",           // +12:00
	"Pacific/Auckland",       // +12:00/+13:00
	"Pacific/Chatham",        // +12:45/+13:45
	"Pacific/Apia",           // +13:00
	"Pacific/Kiritimati",     // +14:00
}

// IsTimezoneValid checks if the given timezone is valid by attempting to load it from the tz database
// This validates against the actual system tz database rather than a hardcoded list
func IsTimezoneValid(tz string) bool {
	if tz == "" {
		return false
	}
	_, err := time.LoadLocation(tz)
	return err == nil
}

// IsCommonTimezone checks if the given timezone is in our curated list of common timezones
func IsCommonTimezone(tz string) bool {
	for _, commonTz := range CommonTimezones {
		if tz == commonTz {
			return true
		}
	}
	return false
}

// GetValidTimezonesString returns a comma-separated string of common timezones for error messages
func GetValidTimezonesString() string {
	return strings.Join(CommonTimezones, ", ")
}

// ConvertTime converts a UTC time to the specified timezone
// Database stores all times in UTC, this function converts them for display
func ConvertTime(utcTime time.Time, targetTimezone string) (time.Time, error) {
	if targetTimezone == "UTC" {
		return utcTime, nil // No conversion needed
	}

	// Load the target timezone location from the tz database
	loc, err := time.LoadLocation(targetTimezone)
	if err != nil {
		return utcTime, fmt.Errorf("failed to load timezone %s: %w", targetTimezone, err)
	}

	// Convert UTC time to the target timezone
	return utcTime.In(loc), nil
}

// GetTimezoneLabel returns a human-readable label for the timezone
// Labels include the offset(s) to make it clear which timezone is which
// Ordered from west to east: -11:00 (Niue) to +14:00 (Kiritimati)
func GetTimezoneLabel(tz string) string {
	// Map of timezone IDs to human-readable labels with offsets
	// Each entry represents a unique STD/DST offset pair
	labels := map[string]string{
		"Pacific/Niue":           "Niue (-11:00)",
		"America/Adak":           "Adak (-10:00/-09:00)",
		"Pacific/Honolulu":       "Honolulu (-10:00)",
		"Pacific/Marquesas":      "Marquesas (-09:30)",
		"America/Anchorage":      "Anchorage (-09:00/-08:00)",
		"Pacific/Gambier":        "Gambier (-09:00)",
		"Pacific/Pitcairn":       "Pitcairn (-08:00)",
		"America/Los_Angeles":    "Los Angeles (-08:00/-07:00)",
		"America/Denver":         "Denver (-07:00/-06:00)",
		"America/Phoenix":        "Phoenix (-07:00)",
		"America/Chicago":        "Chicago (-06:00/-05:00)",
		"America/Mexico_City":    "Mexico City (-06:00)",
		"America/New_York":       "New York (-05:00/-04:00)",
		"America/Lima":           "Lima (-05:00)",
		"America/Barbados":       "Barbados (-04:00)",
		"America/Santiago":       "Santiago (-04:00/-03:00)",
		"America/St_Johns":       "St. John's (-03:30/-02:30)",
		"America/Miquelon":       "Miquelon (-03:00/-02:00)",
		"America/Sao_Paulo":      "São Paulo (-03:00)",
		"America/Godthab":        "Godthab/Nuuk (-02:00/-01:00)",
		"Atlantic/South_Georgia": "South Georgia (-02:00)",
		"Atlantic/Azores":        "Azores (-01:00/+00:00)",
		"Atlantic/Cape_Verde":    "Cape Verde (-01:00)",
		"UTC":                    "UTC (+00:00)",
		"Africa/Abidjan":         "Abidjan (+00:00)",
		"Europe/Dublin":          "Dublin (+00:00/+01:00)",
		"Antarctica/Troll":       "Troll (+00:00/+02:00)",
		"Africa/Lagos":           "Lagos (+01:00)",
		"Europe/Berlin":          "Berlin (+01:00/+02:00)",
		"Africa/Johannesburg":    "Johannesburg (+02:00)",
		"Europe/Athens":          "Athens (+02:00/+03:00)",
		"Africa/Nairobi":         "Nairobi (+03:00)",
		"Asia/Tehran":            "Tehran (+03:30)",
		"Asia/Dubai":             "Dubai (+04:00)",
		"Asia/Kabul":             "Kabul (+04:30)",
		"Asia/Karachi":           "Karachi (+05:00)",
		"Asia/Kolkata":           "Mumbai/Kolkata (+05:30)",
		"Asia/Kathmandu":         "Kathmandu (+05:45)",
		"Asia/Dhaka":             "Dhaka (+06:00)",
		"Asia/Yangon":            "Yangon (+06:30)",
		"Asia/Bangkok":           "Bangkok (+07:00)",
		"Asia/Singapore":         "Singapore (+08:00)",
		"Australia/Eucla":        "Eucla (+08:45)",
		"Asia/Seoul":             "Seoul (+09:00)",
		"Australia/Darwin":       "Darwin (+09:30)",
		"Australia/Adelaide":     "Adelaide (+09:30/+10:30)",
		"Australia/Brisbane":     "Brisbane (+10:00)",
		"Australia/Sydney":       "Sydney (+10:00/+11:00)",
		"Australia/Lord_Howe":    "Lord Howe (+10:30/+11:00)",
		"Pacific/Bougainville":   "Bougainville (+11:00)",
		"Pacific/Norfolk":        "Norfolk (+11:00/+12:00)",
		"Pacific/Fiji":           "Fiji (+12:00)",
		"Pacific/Auckland":       "Auckland (+12:00/+13:00)",
		"Pacific/Chatham":        "Chatham (+12:45/+13:45)",
		"Pacific/Apia":           "Apia (+13:00)",
		"Pacific/Kiritimati":     "Kiritimati (+14:00)",
	}

	if label, exists := labels[tz]; exists {
		return label
	}
	return tz // Return the timezone ID if no label is defined
}

File: web/src/lib/timezone.ts
// src/lib/timezone.ts
// Utility functions for managing timezones in localStorage
// Ordered from west to east: -11:00 (Niue) to +14:00 (Kiritimati)

export type Timezone =
  | 'Pacific/Niue'
  | 'America/Adak'
  | 'Pacific/Honolulu'
  | 'Pacific/Marquesas'
  | 'America/Anchorage'
  | 'Pacific/Gambier'
  | 'America/Los_Angeles'
  | 'Pacific/Pitcairn'
  | 'America/Denver'
  | 'America/Phoenix'
  | 'America/Chicago'
  | 'America/Mexico_City'
  | 'America/New_York'
  | 'America/Lima'
  | 'America/Barbados'
  | 'America/Santiago'
  | 'America/St_Johns'
  | 'America/Miquelon'
  | 'America/Sao_Paulo'
  | 'America/Godthab'
  | 'Atlantic/South_Georgia'
  | 'Atlantic/Azores'
  | 'Atlantic/Cape_Verde'
  | 'UTC'
  | 'Africa/Abidjan'
  | 'Europe/Dublin'
  | 'Antarctica/Troll'
  | 'Africa/Lagos'
  | 'Europe/Berlin'
  | 'Africa/Johannesburg'
  | 'Europe/Athens'
  | 'Africa/Nairobi'
  | 'Asia/Tehran'
  | 'Asia/Dubai'
  | 'Asia/Kabul'
  | 'Asia/Karachi'
  | 'Asia/Kolkata'
  | 'Asia/Kathmandu'
  | 'Asia/Dhaka'
  | 'Asia/Yangon'
  | 'Asia/Bangkok'
  | 'Asia/Singapore'
  | 'Australia/Eucla'
  | 'Asia/Seoul'
  | 'Australia/Darwin'
  | 'Australia/Adelaide'
  | 'Australia/Brisbane'
  | 'Australia/Sydney'
  | 'Australia/Lord_Howe'
  | 'Pacific/Bougainville'
  | 'Pacific/Norfolk'
  | 'Pacific/Fiji'
  | 'Pacific/Auckland'
  | 'Pacific/Chatham'
  | 'Pacific/Apia'
  | 'Pacific/Kiritimati';

const TIMEZONE_STORAGE_KEY = 'velocity-report-timezone';

export function getStoredTimezone(): Timezone | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem(TIMEZONE_STORAGE_KEY) as Timezone | null;
}

export function setStoredTimezone(timezone: Timezone): void {
  if (typeof window === 'undefined') return;
  localStorage.setItem(TIMEZONE_STORAGE_KEY, timezone);
}

export function getDisplayTimezone(defaultTimezone: string): Timezone {
  const stored = getStoredTimezone();
  return stored || (defaultTimezone as Timezone);
}

export function getTimezoneLabel(timezone: Timezone): string {
  switch (timezone) {
    case 'Pacific/Niue':
      return '🇳🇺 Niue (-11:00)';
    case 'America/Adak':
      return '🇺🇸 Adak (-10:00/-09:00)';
    case 'Pacific/Honolulu':
      return '🇺🇸 Honolulu (-10:00)';
    case 'Pacific/Marquesas':
      return '🇵🇫 Marquesas (-09:30)';
    case 'America/Anchorage':
      return '🇺🇸 Anchorage (-09:00/-08:00)';
    case 'Pacific/Gambier':
      return '🇵🇫 Gambier (-09:00)';
    case 'America/Los_Angeles':
      return '🇺🇸 Los Angeles (-08:00/-07:00)';
    case 'Pacific/Pitcairn':
      return '🇵🇳 Pitcairn (-08:00)';
    case 'America/Denver':
      return '🇺🇸 Denver (-07:00/-06:00)';
    case 'America/Phoenix':
      return '🇺🇸 Phoenix (-07:00)';
    case 'America/Chicago':
      return '🇺🇸 Chicago (-06:00/-05:00)';
    case 'America/Mexico_City':
      return '🇲🇽 Mexico City (-06:00)';
    case 'America/New_York':
      return '🇺🇸 New York (-05:00/-04:00)';
    case 'America/Lima':
      return '🇵🇪 Lima (-05:00)';
    case 'America/Barbados':
      return '🇧🇧 Barbados (-04:00)';
    case 'America/Santiago':
      return '🇨🇱 Santiago (-04:00/-03:00)';
    case 'America/St_Johns':
      return '🇨🇦 St. John\'s (-03:30/-02:30)';
    case 'America/Miquelon':
      return '🇵🇲 Miquelon (-03:00/-02:00)';
    case 'America/Sao_Paulo':
      return '🇧🇷 São Paulo (-03:00)';
    case 'America/Godthab':
      return '🇬🇱 Godthab/Nuuk (-02:00/-01:00)';
    case 'Atlantic/South_Georgia':
      return '🇬🇸 South Georgia (-02:00)';
    case 'Atlantic/Azores':
      return '🇵🇹 Azores (-01:00/+00:00)';
    case 'Atlantic/Cape_Verde':
      return '🇨🇻 Cape Verde (-01:00)';
    case 'UTC':
      return '🇺🇳 UTC (+00:00)';
    case 'Africa/Abidjan':
      return '🇨🇮 Abidjan (+00:00)';
    case 'Europe/Dublin':
      return '🇮🇪 Dublin (+00:00/+01:00)';
    case 'Antarctica/Troll':
      return '🇦🇶 Troll (+00:00/+02:00)';
    case 'Africa/Lagos':
      return '🇳🇬 Lagos (+01:00)';
    case 'Europe/Berlin':
      return '🇩🇪 Berlin (+01:00/+02:00)';
    case 'Africa/Johannesburg':
      return '🇿🇦 Johannesburg (+02:00)';
    case 'Europe/Athens':
      return '🇬🇷 Athens (+02:00/+03:00)';
    case 'Africa/Nairobi':
      return '🇰🇪 Nairobi (+03:00)';
    case 'Asia/Tehran':
      return '🇮🇷 Tehran (+03:30)';
    case 'Asia/Dubai':
      return '🇦🇪 Dubai (+04:00)';
    case 'Asia/Kabul':
      return '🇦🇫 Kabul (+04:30)';
    case 'Asia/Karachi':
      return '🇵🇰 Karachi (+05:00)';
    case 'Asia/Kolkata':
      return '🇮🇳 Mumbai (+05:30)';
    case 'Asia/Kathmandu':
      return '🇳🇵 Kathmandu (+05:45)';
    case 'Asia/Dhaka':
      return '🇧🇩 Dhaka (+06:00)';
    case 'Asia/Yangon':
      return '🇲🇲 Yangon (+06:30)';
    case 'Asia/Bangkok':
      return '🇹🇭 Bangkok (+07:00)';
    case 'Asia/Singapore':
      return '🇸🇬 Singapore (+08:00)';
    case 'Australia/Eucla':
      return '🇦🇺 Eucla (+08:45)';
    case 'Asia/Seoul':
      return '🇰🇷 Seoul (+09:00)';
    case 'Australia/Darwin':
      return '🇦🇺 Darwin (+09:30)';
    case 'Australia/Adelaide':
      return '🇦🇺 Adelaide (+09:30/+10:30)';
    case 'Australia/Brisbane':
      return '🇦🇺 Brisbane (+10:00)';
    case 'Australia/Sydney':
      return '🇦🇺 Sydney (+10:00/+11:00)';
    case 'Australia/Lord_Howe':
      return '🇦🇺 Lord Howe (+10:30/+11:00)';
    case 'Pacific/Bougainville':
      return '🇵🇬 Bougainville (+11:00)';
    case 'Pacific/Norfolk':
      return '🇦🇺 Norfolk (+11:00/+12:00)';
    case 'Pacific/Fiji':
      return '🇫🇯 Fiji (+12:00)';
    case 'Pacific/Auckland':
      return '🇳🇿 Auckland (+12:00/+13:00)';
    case 'Pacific/Chatham':
      return '🇳🇿 Chatham (+12:45/+13:45)';
    case 'Pacific/Apia':
      return '🇼🇸 Apia (+13:00)';
    case 'Pacific/Kiritimati':
      return '🇰🇮 Kiritimati (+14:00)';
    default:
      return timezone;
  }
}

export const AVAILABLE_TIMEZONES: { value: Timezone; label: string }[] = [
  { value: 'Pacific/Niue', label: '🇳🇺 Niue (-11:00)' },
  { value: 'America/Adak', label: '🇺🇸 Adak (-10:00/-09:00)' },
  { value: 'Pacific/Honolulu', label: '🇺🇸 Honolulu (-10:00)' },
  { value: 'Pacific/Marquesas', label: '🇵🇫 Marquesas (-09:30)' },
  { value: 'America/Anchorage', label: '🇺🇸 Anchorage (-09:00/-08:00)' },
  { value: 'Pacific/Gambier', label: '🇵🇫 Gambier (-09:00)' },
  { value: 'America/Los_Angeles', label: '🇺🇸 Los Angeles (-08:00/-07:00)' },
  { value: 'Pacific/Pitcairn', label: '🇵🇳 Pitcairn (-08:00)' },
  { value: 'America/Denver', label: '🇺🇸 Denver (-07:00/-06:00)' },
  { value: 'America/Phoenix', label: '🇺🇸 Phoenix (-07:00)' },
  { value: 'America/Chicago', label: '🇺🇸 Chicago (-06:00/-05:00)' },
  { value: 'America/Mexico_City', label: '🇲🇽 Mexico City (-06:00)' },
  { value: 'America/New_York', label: '🇺🇸 New York (-05:00/-04:00)' },
  { value: 'America/Lima', label: '🇵🇪 Lima (-05:00)' },
  { value: 'America/Barbados', label: '🇧🇧 Barbados (-04:00)' },
  { value: 'America/Santiago', label: '🇨🇱 Santiago (-04:00/-03:00)' },
  { value: 'America/St_Johns', label: '🇨🇦 St. John\'s (-03:30/-02:30)' },
  { value: 'America/Miquelon', label: '🇵🇲 Miquelon (-03:00/-02:00)' },
  { value: 'America/Sao_Paulo', label: '🇧🇷 São Paulo (-03:00)' },
  { value: 'America/Godthab', label: '🇬🇱 Godthab/Nuuk (-02:00/-01:00)' },
  { value: 'Atlantic/South_Georgia', label: '🇬🇸 South Georgia (-02:00)' },
  { value: 'Atlantic/Azores', label: '🇵🇹 Azores (-01:00/+00:00)' },
  { value: 'Atlantic/Cape_Verde', label: '🇨🇻 Cape Verde (-01:00)' },
  { value: 'UTC', label: '🇺🇳 UTC (+00:00)' },
  { value: 'Africa/Abidjan', label: '🇨🇮 Abidjan (+00:00)' },
  { value: 'Europe/Dublin', label: '🇮🇪 Dublin (+00:00/+01:00)' },
  { value: 'Antarctica/Troll', label: '🇦🇶 Troll (+00:00/+02:00)' },
  { value: 'Africa/Lagos', label: '🇳🇬 Lagos (+01:00)' },
  { value: 'Europe/Berlin', label: '🇩🇪 Berlin (+01:00/+02:00)' },
  { value: 'Africa/Johannesburg', label: '🇿🇦 Johannesburg (+02:00)' },
  { value: 'Europe/Athens', label: '🇬🇷 Athens (+02:00/+03:00)' },
  { value: 'Africa/Nairobi', label: '🇰🇪 Nairobi (+03:00)' },
  { value: 'Asia/Tehran', label: '🇮🇷 Tehran (+03:30)' },
  { value: 'Asia/Dubai', label: '🇦🇪 Dubai (+04:00)' },
  { value: 'Asia/Kabul', label: '🇦🇫 Kabul (+04:30)' },
  { value: 'Asia/Karachi', label: '🇵🇰 Karachi (+05:00)' },
  { value: 'Asia/Kolkata', label: '🇮🇳 Mumbai (+05:30)' },
  { value: 'Asia/Kathmandu', label: '🇳🇵 Kathmandu (+05:45)' },
  { value: 'Asia/Dhaka', label: '🇧🇩 Dhaka (+06:00)' },
  { value: 'Asia/Yangon', label: '🇲🇲 Yangon (+06:30)' },
  { value: 'Asia/Bangkok', label: '🇹🇭 Bangkok (+07:00)' },
  { value: 'Asia/Singapore', label: '🇸🇬 Singapore (+08:00)' },
  { value: 'Australia/Eucla', label: '🇦🇺 Eucla (+08:45)' },
  { value: 'Asia/Seoul', label: '🇰🇷 Seoul (+09:00)' },
  { value: 'Australia/Darwin', label: '🇦🇺 Darwin (+09:30)' },
  { value: 'Australia/Adelaide', label: '🇦🇺 Adelaide (+09:30/+10:30)' },
  { value: 'Australia/Brisbane', label: '🇦🇺 Brisbane (+10:00)' },
  { value: 'Australia/Sydney', label: '🇦🇺 Sydney (+10:00/+11:00)' },
  { value: 'Australia/Lord_Howe', label: '🇦🇺 Lord Howe (+10:30/+11:00)' },
  { value: 'Pacific/Bougainville', label: '🇵🇬 Bougainville (+11:00)' },
  { value: 'Pacific/Norfolk', label: '🇦🇺 Norfolk (+11:00/+12:00)' },
  { value: 'Pacific/Fiji', label: '🇫🇯 Fiji (+12:00)' },
  { value: 'Pacific/Auckland', label: '🇳🇿 Auckland (+12:00/+13:00)' },
  { value: 'Pacific/Chatham', label: '🇳🇿 Chatham (+12:45/+13:45)' },
  { value: 'Pacific/Apia', label: '🇼🇸 Apia (+13:00)' },
  { value: 'Pacific/Kiritimati', label: '🇰🇮 Kiritimati (+14:00)' }
];
